### PR TITLE
rand: add callbacks to cleanup the user entropy resp. nonce

### PR DESCRIPTION
**Pull request by mspncp**
The `get_user_{entropy,nonce}` callbacks were add recently to the dispatch table in commit 4cde7585ce8e. Instead of adding corresponding `cleanup_user_{entropy,nonce}` callbacks, the `cleanup_{entropy,nonce}` callbacks were reused. This can cause a problem in the case where the seed source is replaced by a provider: the buffer gets allocated by the provider but cleared by the core.

### CODE REVIEW

File: crypto/rand/prov_seed.c
Comment (mspncp) [inline crypto/rand/prov_seed.c:144]: Without the changes from #22394, the implementations  of  `ossl_rand_cleanup_user_nonce` and `ossl_rand_cleanup_nonce`  are identical. It's important however, that both callbacks exist.

Now  #22394 will depend on this pull request to be merged first.

**Review by t8m — CHANGES_REQUESTED** [review]
One small doc fix is needed, other than that LGTM.

File: doc/man7/provider-base.pod
Comment (mspncp) [reply doc/man7/provider-base.pod:329]: I just noticed that I forgot to push the latest version of my branch. Let me push it first (it's a plain rebase on master) and then I'll address your request.

grafik

Comment (mspncp) [reply doc/man7/provider-base.pod:329]: done

**Comment by t8m** [comment]
Ping for second review. This needs to go in before the release of 3.1.4.

**Comment by mattcaswell** [comment]
Pushed to master and 3.1

**Comment by paulidale** [comment]
I don't understand the problem here.  Well I do understand the issue but am not sure if it's possible to encounter it in reality.

**Comment by mspncp** [comment]
I would expect problems on Windows if the provider links to a different crt library than libcrypto. 

On Linux it's probably not such a big problem iiuc, because all binaries tend to use the same libc.

**Comment by paulidale** [comment]
I'm more thinking along the lines of the provider replacing the seed src will implement correctly paired functions.  If the new callbacks are available, the old ones will not be used so the clearing and freeing will be consistent.  Likewise if only the old are implemented.  Getting a mismatch seems tricky.

**Comment by paulidale** [comment]
I don't think this was required.  The memory in all cases is being allocated and freed by libcrypto, not a provider.  This is true even if the provider implements its own seed source.

These functions are callbacks from the provider to the core libcrypto, that they then might call into a provider is moot because the allocations happen before then.

**Comment by paulidale** [comment]
With #22394, this becomes necessary however.

**Pull request closed without merging**
